src/webapi/selection.rs

```diff
@@ -0,0 +1,341 @@
+    /// is to say that the selection's start and end points are at the same position in the
+    /// content.
+    ///
+    /// Keep in mind that a collapsed selection may still have one (or more, in Gecko)
+    /// [Range](struct.Range.html)s, so [range_count](struct.Selection.html#method.range_count) may
+    /// not be zero. In that scenario, calling a [Selection](struct.Selection.html) object's
+    /// [get_range_at](struct.Selection.html#method.get_range_at) method may return a
+    /// [Range](struct.Range.html) object which is collapsed.
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Selection/isCollapsed)
+    pub fn is_collapsed(&self) -> bool {
+        js! (
+            return @{self}.isCollapsed;
+        ).try_into().unwrap()
+    }
+
+    /// Returns the number of ranges in the selection.
+    ///
+    /// Before the user has clicked a freshly loaded page, the
+    /// [range_count](struct.Selection.html#method.range_count) is 0. After the user
+    /// clicks on the page, [range_count](struct.Selection.html#method.range_count) even if no 
+    /// selection is visible.
+    ///
+    /// A user can normally only select one range at a time, so the 
+    /// [range_count](struct.Selection.html#method.range_count) will usually be 1.
+    /// Scripting can be used to make the selection contain more than 1 range.
+    ///
+    /// Gecko browsers allow multiple selections across table cells.
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Selection/rangeCount)
+    pub fn range_count(&self) -> u32 {
+        js! (
+            return @{self}.rangeCount;
+        ).try_into().unwrap()
+    }
+
+    /// Returns the type of the current selection.
+    ///
+    /// Possible values are:
+    ///
+    /// * `None`: No selection has currently been made.
+    /// * `Caret`: The selection is collapsed (i.e. the caret is placed on some text, but no
+    /// range has been selected).
+    /// * `Range`: A range has been selected.
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Selection/type)
+    pub fn selection_type(&self) -> String {
+        js! (
+            return @{self}.type;
+        ).try_into().unwrap()
+    }
+
+    /// Returns a [Range](struct.Range.html) object representing one of the ranges currently
+    /// selected.
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Selection/getRangeAt)
+    pub fn get_range_at(&self, index: u32) -> Range {
+        js! (
+            return @{self}.getRangeAt(@{index});
+        ).try_into().unwrap()
+    }
+
+    /// Adds a [Range](struct.Range.html) to the [Selection](struct.Selection.html).
+    ///
+    /// [(Javascript docs)](https://developer.mozilla.org/en-US/docs/Web/API/Selection/addRange)
+    pub fn add_range(&self, range: Range) -> Result<(), TODO> {
```

Comment:
1. Why does it return `Result`? Can it throw? The specs say only that it can "abort these steps". Can you check and make sure?
2. Pass the range by reference: `&Range`.